File: 2019/02/solution.py
from operator import add, mul


def parse_input():
    with open('input.txt') as f:
        line = f.read()
    return list(map(int, line.split(',')))

def part1(intcode):
    return run_intcode(intcode, 12, 2)

def part2(intcode):
    for noun in range(100):
        for verb in range(100):
            if run_intcode(intcode, noun, verb) == 19690720:
                return 100 * noun + verb

def run_intcode(intcode, noun, verb):
    memory = intcode.copy()
    memory[1] = noun
    memory[2] = verb
    ip = 0

    def operate(op):
        memory[memory[ip + 3]] = op(
            memory[memory[ip + 1]],
            memory[memory[ip + 2]],
        )

    while True:
        if memory[ip] == 1:
            operate(add)
        elif memory[ip] == 2:
            operate(mul)
        elif memory[ip] == 99:
            return memory[0]
        else:
            raise Exception(f'Unknown opcode {memory[ip]}')
        ip += 4


intcode = parse_input()
print(part1(intcode))
print(part2(intcode))

File: 2019/02/solution.rs
use std::fs;

fn main() {
    let input = parse_input();
    println!("{}", part1(&input));
    println!("{}", part2(&input).unwrap());
}

fn parse_input() -> Vec<i32> {
    fs::read_to_string("input.txt")
        .unwrap()
        .split(",")
        .map(|number| number.parse().unwrap())
        .collect()
}

fn part1(input: &Vec<i32>) -> i32 {
    run_intcode(input, 12, 2)
}

fn part2(input: &Vec<i32>) -> Option<i32> {
    for noun in 1..100 {
        for verb in 1..100 {
            if run_intcode(input, noun, verb) == 19690720 {
                return Some(100 * noun + verb);
            }
        }
    }
    return None;
}

fn run_intcode(intcode: &Vec<i32>, noun: i32, verb: i32) -> i32 {
    let mut memory = intcode.clone();
    memory[1] = noun;
    memory[2] = verb;
    let mut ip: usize = 0;

    loop {
        match memory[ip] {
            1 => operate(&mut memory, &ip, |x, y| x + y),
            2 => operate(&mut memory, &ip, |x, y| x * y),
            99 => break,
            _ => panic!("Unknown operation"),
        }
        ip += 4;
    }
    memory[0]
}

fn operate(memory: &mut Vec<i32>, ip: &usize, op: fn(i32, i32) -> i32) {
    let params: Vec<usize> = (1..4_usize).map(|i| memory[ip + i] as usize).collect();
    memory[params[2]] = op(memory[params[0]], memory[params[1]]);
}
